stdlib/public/Observation/Sources/Observation/OptimizedSet.swift

```diff
@@ -0,0 +1,123 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+// Observation favors sets with either 1 or 0 elements; this prevents some of 
+// the extra traffic of retains/releases and hashes by specializing the
+// storage to a enumeration for holding those values and falling back to a 
+// standard Set.
+
+enum OptimizedSet<Element: Hashable> {
+	case empty
+	case single(Element)
+	case many(Set<Element>)
+}
+
+extension OptimizedSet {
+	var isEmpty: Bool {
+		switch self {
+		case .empty: return true
+		case .single: return false
+		case .many(let elements): return elements.isEmpty
+		}
+	}
+
+	mutating func insert(_ element: Element) {
+		switch self {
+		case .empty: 
+			self = .single(element)
+		case .single(let existing):
+			guard element != existing else {
+				return
+			}
+			self = .many([existing, element])
+		case .many(var existing):
+			self = .empty // avoid CoW
+			existing.insert(element)
+			self = .many(existing)
+		}
+	}
+
+	mutating func remove(_ element: Element) {
+		switch self {
+		case .empty:
+			break
+		case .single(let existing):
+			if existing == element {
+				self = .empty
+			}
+		case .many(var existing):
+			self = .empty
+			existing.remove(element)
+			let count = existing.count
+			if count == 1 {
+				self = .single(existing.first!)
+			} else {
+				self = .many(existing)
```

Comment:
there's an implied invariant here that `count > 1` right? is that worth enforcing or documenting somewhere?